### PR TITLE
cache hashes of accesspaths; keep exclusions of accesspaths sorted

### PR DESCRIPTION
**Pull request by bbrehm**
The caching of accesspath hashes gives > 5% speedup on dataflow heavy loads (cumulative to https://github.com/ShiftLeftSecurity/overflowdb/pull/194).

It is also saves some expansions (more context merges) in the closed source tracker; and it is a useful preparation for OVERRIDE/PRESERVE handling of policies.

### CODE REVIEW

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/accesspath/AccessPath.scala [hunk 1]
```diff
@@ -49,9 +51,10 @@ object AccessPath {
       val nElements = el.elements.length - 1
       while (nElements - i > -1) {
         el.elements(nElements - i) match {
-          case (AddressOf | VariablePointerShift | _: PointerShift) => i += 1
+          case (AddressOf | VariablePointerShift | _: PointerShift) =>
           case _                                                    => return i
         }
+        i += 1
```
Comment (ml86) [inline semanticcpg/src/main/scala/io/shiftleft/semanticcpg/accesspath/AccessPath.scala:57]: Why?

**Review by ml86 — APPROVED** [review]

File: semanticcpg/src/main/scala/io/shiftleft/semanticcpg/accesspath/AccessPath.scala (hunk 1 above)
Comment (fabsx00) [reply semanticcpg/src/main/scala/io/shiftleft/semanticcpg/accesspath/AccessPath.scala:57]: Do we need this unaesthetic while loop here?

Comment (bbrehm) [reply semanticcpg/src/main/scala/io/shiftleft/semanticcpg/accesspath/AccessPath.scala:57]: This specific change is purely for readability -- it does not change the control flow graph, and if the scala compiler does its job then it should not even change the bytecode.

Regarding whether we need the loop at all -- this is hot code, and I see no formulation that is both more obvious in terms of semantics and more obvious in terms of performance (i.e. in terms of what we hope for the style of machine code).

Comment (ml86) [reply semanticcpg/src/main/scala/io/shiftleft/semanticcpg/accesspath/AccessPath.scala:57]: To me the old version was the better one.

**Pull request merged**